packages/vite/src/node/plugins/worker.ts

```diff
@@ -241,15 +270,48 @@ export function webWorkerPlugin(config: ResolvedConfig): Plugin {
         code: `export default function WorkerWrapper() {
           return new ${workerConstructor}(${JSON.stringify(
           url
-        )}, ${JSON.stringify(workerOptions, null, 2)})
+        )}, ${JSON.stringify(workerOptions)})
         }`,
         map: { mappings: '' } // Empty sourcemap to suppress Rollup warning
       }
     },
 
-    renderChunk(code) {
-      if (config.isWorker && code.includes('import.meta.url')) {
-        return code.replace('import.meta.url', 'self.location.href')
+    renderChunk(code, chunk) {
+      let s: MagicString
+      const result = () => {
+        return (
+          s && {
+            code: s.toString(),
+            map: config.build.sourcemap ? s.generateMap({ hires: true }) : null
+          }
+        )
+      }
+      if (code.match(workerAssetUrlRE) || code.includes('import.meta.url')) {
+        let match: RegExpExecArray | null
+        s = new MagicString(code)
+
+        // Replace "__VITE_WORKER_ASSET__5aa0ddc0__" using relative paths
+        const workerMap = workerCache.get(config.mainConfig || config)!
+        const { fileNameHash } = workerMap
+
+        while ((match = workerAssetUrlRE.exec(code))) {
+          const [full, hash] = match
+          const filename = fileNameHash.get(hash)!
+          const outputFilepath = path.relative(
+            path.dirname(chunk.fileName),
+            filename
+          )
+          const replacement = JSON.stringify(outputFilepath).slice(1, -1)
+          s.overwrite(match.index, match.index + full.length, replacement, {
+            contentOnly: true
+          })
+        }
+
+        // TODO: check if this should be removed
+        if (config.isWorker) {
+          s = s.replace('import.meta.url', 'self.location.href')
+          return result()
+        }
```

Comment:
use `resolveImportMeta` hook to do it. like this, https://github.com/vitejs/vite/pull/7837/files#diff-3b7bb11b5633d82f82c069037789e105005d657667b1da1e96c18fcf3eb622adL288-L292. I think we should open a new PR to resolve it

Reply:
Nice! Yes, let's review the handling of `import.meta.url` in a separate PR. This one is too big already. Better to try to merge it so we can keep building on top of it.